README.md
# password-policy

My own password policy for apps I'm creating implemented in JavaScript. 

## Documentation

Documentation can be found here: [password-policy documentation](https://dzek69.github.io/password-policy).
Jump directly to usage information: [password-policy usage](https://dzek69.github.io/password-policy/tutorial-EXAMPLE.html).

## Demo

Demo can be found here: [password-policy demo](https://dzek69.github.io/password-policy/demo).

## Features

- returns valid flag, "security level" in percents (for nice progress bars under password inputs) and scored points
(for demo purposes, but you can save it with the password too, so if you wanna enforce better passwords in the future
you can do it)
- tweakable, but by default it should:
    - filter out most weak passwords,
    - prevent reusing same characters,
    - be BS-free and requiring just unique characters, not different character ranges (lower/upper case letters,
    numbers, some "special chars" and that kind of typical BS) 
- es6+ first approach, >3% browsers usage transpiled version to be found inside `dist` folder *

\* - transpiling kills JS engines optimizations, makes codes longer and tree shaking harder to do and/or slower

## Should I use it?

It's your choice if you like the idea. I published this for my own needs. Still would like to hear your thoughts/ideas.

## To be done

- Unit tests ?

## License

MIT


CHANGELOG.md
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [UNRELEASED]
(nothing yet)

## [2.0.0] - 2019-08-31
### Changed
- `dist` transpiling settings
### Fixed
- dev deps audit warnings

## [1.0.0] - 2018-05-31
### Added
- first version


src/index.spec.js
"use strict";

const validator = require("./index");

const log = (value) => {
    console.info(value, validator(value));
};

log("Hi");
log("123456");
log("aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa");
log("HelloThere");
log("My password is not weak");
log("Correct horse battery staple");


src/index.js
"use strict";

/**
 * Creates new validator with some rules overriden. Use `tweaked.js` export to more friendly configuration (object with
 * defaults, instead of long list of required arguments)
 * @param {number} fullPoints - points required to score for value to be marked as "100% secure"
 * @param {number} requiredPercentage - percentage value required for password to be considered safe
 * @param {number} pointsPerChar - how many points are added per character
 * @param {number} pointsPerUnique - how many additional points are added per unique character
 * @param {number} repeatedCharPenalty - how many points are taken for repeated characters
 * @param {number} repeatedCharCount - how many times a character should be repeated before being considered
 * "repeated"
 * @param {number} repeatedCharMaxPosition - position in string after characters are no longer considered
 * repeated (this is used to avoid longer passwords to be considered less and less secure, because user run out of
 * unique characters on keyboard)
 * @param {number} minUniqueChars - minimal count of unique characters required to not lowering the points to
 * `maxNonUniqueScore` value
 * @param {number} maxNonUniqueScore - if minimal count of unique characters is not met - this is the point to
 * the score will be lowered (only if already higher)
 * @returns {Validator}
 */
const createValidator = (
    fullPoints, requiredPercentage, pointsPerChar, pointsPerUnique, repeatedCharPenalty, repeatedCharCount,
    repeatedCharMaxPosition, minUniqueChars, maxNonUniqueScore,
) => {
    const calculate = (value) => {
        const chars = value.split("");
        const map = {};
        let points = 0;

        chars.forEach((char, index) => {
            if (!map[char]) { map[char] = 0; }
            map[char]++;

            points += pointsPerChar; // every new char

            if (map[char] === 1) { // unique char
                points += pointsPerUnique;
            }
            if (map[char] > repeatedCharCount) { // non unique char
                if (index <= repeatedCharMaxPosition) { // for char on position <= 12
                    points -= repeatedCharPenalty;
                }
            }
        });

        if (Object.keys(map).length <= minUniqueChars) {
            if (points > maxNonUniqueScore) {
                points = maxNonUniqueScore;
            }
        }

        return points;
    };

    const format = (points) => {
        const percentage = Math.min(points * (100 / fullPoints), 100); // eslint-disable-line no-magic-numbers
        const valid = percentage >= requiredPercentage;
        return {
            points,
            percentage,
            valid,
        };
    };

    return value => format(calculate(value));
};

/**
 * @typedef {Object} Result
 * @property {number} points - scored points
 * @property {number} percentage - percentage value of password security (0 - 100)
 * @property {boolean} valid - is password score considered secure enough
 */

/**
 * @typedef {function} Validator
 * @param {string} value - password to measure its security level
 * @returns {Result}
 */

/**
 * @type {Validator}
 */
const validate = createValidator(18, 75, 1, 0.5, 1.5, 3, 7, 5, 5); // eslint-disable-line no-magic-numbers
validate.createValidator = createValidator;

module.exports = validate;


src/tweaked.js
"use strict";

const { createValidator } = require("./index");

const defaults = [18, 75, 1, 0.5, 1.5, 3, 7, 5, 5]; // eslint-disable-line no-magic-numbers
const props = [
    "fullPoints", "requiredPercentage", "pointsPerChar", "pointsPerUnique", "repeatedCharPenalty", "repeatedCharCount",
    "repeatedCharMaxPosition", "minUniqueChars", "maxNonUniqueScore",
];

const createOptions = object => {
    const options = {};
    props.forEach(prop => {
        if (prop in object) {
            options[prop] = object[prop];
            return;
        }
        const index = props.indexOf(prop);
        options[prop] = defaults[index];
    });
    return options;
};

const unpackOptions = options => {
    return props.map(prop => options[prop]);
};

/**
 * @typedef {Object} TweakedOptions
 * @property {number} [fullPoints = 18] - points required to score for value to be marked as "100% secure"
 * @property {number} [requiredPercentage = 75] - percentage value required for password to be considered safe
 * @property {number} [pointsPerChar = 1] - how many points are added per character
 * @property {number} [pointsPerUnique = 0.5] - how many additional points are added per unique character
 * @property {number} [repeatedCharPenalty = 1.5] - how many points are taken for repeated characters
 * @property {number} [repeatedCharCount = 3] - how many times a character should be repeated before being considered
 * "repeated"
 * @property {number} [repeatedCharMaxPosition = 7] - position in string after characters are no longer considered
 * repeated (this is used to avoid longer passwords to be considered less and less secure, because user run out of
 * unique characters on keyboard)
 * @property {number} [minUniqueChars = 5] - minimal count of unique characters required to not lowering the points to
 * `maxNonUniqueScore` value
 * @property {number} [maxNonUniqueScore = 5] - if minimal count of unique characters is not met - this is the point to
 * the score will be lowered (only if already higher)
 */

/**
 * Creates tweaked validator.
 *
 * @param {TweakedOptions} options
 * @returns {Validator} - validator function
 */

const createTweakedValidator = options => {
    return createValidator(...unpackOptions(createOptions(options)));
};
module.exports = createTweakedValidator;
